### PR TITLE
New Python Support Framework

### PR DESCRIPTION
**Pull request by AlexCatarino**
First commit with changes related to the new python support framework based on pythonnet

### CODE REVIEW

**Comment by AlexCatarino** [comment]
I have edit the readme file to inform we need to change the directories where the compiler look for the pythonnet dll (**Python.Runtime.dll**):
`sed -i -e 's/UCS2/UCS4/' Algorithm/QuantConnect.Algorithm.csproj`
`sed -i -e 's/UCS2/UCS4/' AlgorithmFactory/QuantConnect.AlgorithmFactory.csproj`
However we need to provide the dll mapping file (**Python.Runtime.dll.config**) that needs to be placed in the executable folder together with the pythonnet dll. There are two options:
1. Include this file in Lean
2. Download it from github:
`wget https://raw.githubusercontent.com/QuantConnect/pythonnet/master/Python.Runtime.dll.config`

File: Algorithm/QuantConnect.Algorithm.csproj [hunk 1]
```diff
@@ -72,6 +72,10 @@
       <SpecificVersion>False</SpecificVersion>
       <HintPath>..\packages\NodaTime.1.3.1\lib\net35-Client\NodaTime.dll</HintPath>
     </Reference>
+    <Reference Include="Python.Runtime, Version=4.0.0.1, Culture=neutral, processorArchitecture=AMD64">
+      <HintPath>..\packages\AlexCatarino.PythonNET.Runtime.1.0.9\lib\UCS2\Python.Runtime.dll</HintPath>
```
Comment (jaredbroad) [inline Algorithm/QuantConnect.Algorithm.csproj]: Please tidy this up to reference the QuantConnect Nuget package.

File: Algorithm/QuantConnect.Algorithm.csproj [hunk 2]
```diff
@@ -115,6 +124,13 @@
   </ItemGroup>
   <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
   <Import Project="$(SolutionDir)\.nuget\NuGet.targets" Condition="Exists('$(SolutionDir)\.nuget\NuGet.targets')" />
+  <Import Project="..\packages\AlexCatarino.PythonNET.Runtime.1.0.9\build\AlexCatarino.PythonNET.Runtime.targets" Condition="Exists('..\packages\AlexCatarino.PythonNET.Runtime.1.0.9\build\AlexCatarino.PythonNET.Runtime.targets')" />
+  <Target Name="EnsureNuGetPackageBuildImports" BeforeTargets="PrepareForBuild">
+    <PropertyGroup>
+      <ErrorText>This project references NuGet package(s) that are missing on this computer. Use NuGet Package Restore to download them.  For more information, see http://go.microsoft.com/fwlink/?LinkID=322105. The missing file is {0}.</ErrorText>
+    </PropertyGroup>
+    <Error Condition="!Exists('..\packages\AlexCatarino.PythonNET.Runtime.1.0.9\build\AlexCatarino.PythonNET.Runtime.targets')" Text="$([System.String]::Format('$(ErrorText)', '..\packages\AlexCatarino.PythonNET.Runtime.1.0.9\build\AlexCatarino.PythonNET.Runtime.targets'))" />
```
Comment (jaredbroad) [inline Algorithm/QuantConnect.Algorithm.csproj]: Same with this change block; make sure to reference the official QC-Pythonnet fork not your personal repo.

File: AlgorithmFactory/Loader.cs [hunk 3]
```diff
@@ -140,80 +140,88 @@ public bool TryCreateAlgorithmInstance(string assemblyPath, out IAlgorithm algor
...
         private bool TryCreatePythonAlgorithm(string assemblyPath, out IAlgorithm algorithmInstance, out string errorMessage)
         {
-            var success = false;
             algorithmInstance = null;
             errorMessage = "";
 
+            var algorithmName = Config.Get("algorithm-type-name", "main");
+
+            // Copies file to execution location
+            foreach (var file in new DirectoryInfo(@"../../../Algorithm.Python/").GetFiles("*.py"))
```
Comment (jaredbroad) [inline AlgorithmFactory/Loader.cs]: Please make the python file location loaded from a config setting.

File: AlgorithmFactory/Loader.cs [hunk 4]
```diff
@@ -140,80 +140,88 @@ public bool TryCreateAlgorithmInstance(string assemblyPath, out IAlgorithm algor
...
+                    
+                    foreach (var name in scope.Dir())
+                    {
+                        var attr = scope.GetAttr(name.ToString());
+
+                        if (attr.IsSubclass(baseClass) && attr.Repr().Contains(algorithmName))
                         {
-                            dynamicAlgorithm = item.Value;
+                            var addUniverse = Py.Import("qc_algorithm_helper").GetAttr("AddUniverse");
+                            attr.SetAttr("Add_Universe", addUniverse);
```
Comment (jaredbroad) [inline AlgorithmFactory/Loader.cs]: If anyone has ideas how to remove this we'd like to ideally keep the "AddUniverse" method name but override its implementation with our own helper method. This underscore is a temporary work around but will cause confusion if its pushed to production.

File: AlgorithmFactory/QuantConnect.AlgorithmFactory.csproj [hunk 5]
```diff
@@ -80,6 +84,10 @@
     <Compile Include="Properties\AssemblyInfo.cs" />
   </ItemGroup>
   <ItemGroup>
+    <ProjectReference Include="..\Algorithm\QuantConnect.Algorithm.csproj">
+      <Project>{3240ACA4-BDD4-4D24-AC36-BBB651C39212}</Project>
+      <Name>QuantConnect.Algorithm</Name>
```
Comment (jaredbroad) [inline AlgorithmFactory/QuantConnect.AlgorithmFactory.csproj]: Why did you need to add the algorithm to the algorithm factory project? Please also make sure to only reference the QuantConnect fork of pythonnet.

File: Algorithm/QCAlgorithm.cs [hunk 6]
```diff
@@ -1558,5 +1558,42 @@ private Security CreateBenchmarkSecurity()
...
+
+        /// <summary>
+        /// Set the runtime error
+        /// </summary>
+        /// <param name="exception">Represents error that occur during execution</param>
+        public void SetRunTimeError(Exception exception)
+        {
+            if (exception == null)
+            {
+                throw new ArgumentNullException("exception");
```
Comment (jaredbroad) [inline Algorithm/QCAlgorithm.cs]: Please put a better error message here, or none at all if possible. Same with history provider above.

File: Algorithm/QuantConnect.Algorithm.csproj (hunk 1 above)
Comment (AlexCatarino) [inline Algorithm/QuantConnect.Algorithm.csproj]: I created this nuget package with my personal account in nuget. I think QC should create its own package in its own account.

File: AlgorithmFactory/QuantConnect.AlgorithmFactory.csproj (hunk 5 above)
Comment (AlexCatarino) [inline AlgorithmFactory/QuantConnect.AlgorithmFactory.csproj]: In Loader.cs, we call the python wrapper that belongs to the QuantConnect.Algorithm project.
